Add `code_files` and `test_files` options

Genki-S opened the pull request:
Implementation of an idea discussed in #5 

In order to make it possible to narrow the target of test and code files, I implemented the idea of providing `code_files` option discussed in #5 .

Usage is like this:

```
TTNT::TestTask.new do |t|
  t.test_files = FileList['**/*_test.rb']
  t.code_files = FileList['lib/**/*.rb'] - FileList['lib/vendor/**/*.rb']
end
```

I took an approach to filter generated mapping after `anchor` task is finished. Filtering method is added by [this commit (Implement `TestToCodeMapping#select_code_files!)](https://github.com/Genki-S/ttnt/commit/fb47020b38aebc0b0adccb2edae4cc10bbfc905d) and the use of the method is added in [this commit (Filter mapping based on code_files option after anchoring)](https://github.com/Genki-S/ttnt/commit/2c55d988f06d16994d5a1c58fa276aa776b796ab).

@robin850 could you review the code please?

robin850 commented on test/testtask_test.rb:
```diff
@@ -32,4 +32,21 @@ def test_expanded_file_list
+  def test_instance_without_passing_rake_task
+    default_rake_task = Rake::TestTask.new
+    ttnt_task = TTNT::TestTask.new
+    assert_equal default_rake_task.name, ttnt_task.instance_variable_get(:@rake_testtask).name
```
What do you think about:

~~~ruby
assert ttnt_task.instance_variable_get(:@rake_testtask).kind_of?(Rake::TestTask)
~~~

Genki-S replied:
That's much better, thank you!
Fixed here: https://github.com/Genki-S/ttnt/commit/d4d8373e2cff663d5235bf8bdac2914071a394e2

robin850 commented on test/integration_test.rb:
```diff
@@ -54,5 +54,20 @@ def test_selecting_test_even_though_addition_is_made_among_comments
+        File.write(fn, File.read(fn).gsub(/"fizzbuzz"$/, "foo"))
+        output = rake('ttnt:test:run')
+        assert_match 'No test selected.', output[:stderr],
+          'Changing files which is not specified in code_files should not select tests.'
```
"which *are* not ... should not select *any* tests."

Genki-S replied:
Oops... thank you!
Fixed: https://github.com/Genki-S/ttnt/commit/77ae19197d8fe6c6da37aa45cdfe2990b1d1b1ad

robin850 commented on lib/ttnt/testtask.rb:
```diff
@@ -12,14 +13,21 @@ class TestTask
     attr_accessor :rake_testtask
 
+    attr_accessor :test_files
+    attr_accessor :code_files
```
What do you think about doing:

~~~ruby
attr_reader :test_files
attr_reader :code_files

def code_files=(files)
  @code_files = files.kind_of?(String) ? FileList[files] : files
end

def test_files=(files)
  @test_files = files.kind_of?(String) ? FileList[files] : files
end
~~~

Genki-S replied:
This is great! Thank you!
Did it here: https://github.com/Genki-S/ttnt/commit/173fa9795b7ae67ed895900fbc7824b523d90428

robin850 commented:
Again you can maybe update the README file documenting this feature ; I would also update the old syntax, something along the lines of:

> ### Define Rake tasks
>
> TTNT allows you to define its tasks according to an existing `Rake::TestTask` object like:
>
> ~~~ruby
> require 'rake/testtask'
> require 'ttnt/testtask'
>
> t = Rake::TestTask.new do |t|
>   t.libs << "test"
>   t.name = 'task_name'
> end
>
> TTNT::TestTask.new(t)
> ~~~
>
> This will define 2 tasks: `ttnt:task_name:anchor` and `ttnt:task_name:run`. Usage for those tasks are described later in this document.
>
> You can also instantiate a new `TTNT::TestTask` object and specify certain options like:
>
> ~~~ruby
> require 'ttnt/testtask'
>
> TTNT::TestTask.new do |t|
>   t.code_files = FileList['lib/**/*.rb'] - FileList['lib/vendor/**/*.rb']
>   t.test_files = "test/**/*_test.rb"
> end
> ~~~

Also, what do you think about providing a name option in case we are not passing any `Rake::TestTask` object ?

robin850 commented:
Also I guess you should update `TTNT::TestToCodeMapping#select_project_files` to check whether it's included inside `code_files` rather than if it starts with the repository's root.

Genki-S commented:
> Again you can maybe update the README file documenting this feature

Oops... I forgot again :disappointed: Thank you for even writing the content! I updated README: https://github.com/Genki-S/ttnt/commit/8ae206df3c6745f8bcbec53840b3df467899920b

> Also, what do you think about providing a name option in case we are not passing any Rake::TestTask object 

That's nice. I added it here: https://github.com/Genki-S/ttnt/commit/b3c0370c44a0b3a0dfd7698044432de6b84ba2a3

> Also I guess you should update TTNT::TestToCodeMapping#select_project_files to check whether it's included inside code_files rather than if it starts with the repository's root.

I'm dealing with this by filtering generated mapping after `anchor` task is finished. [These lines](https://github.com/Genki-S/ttnt/blob/code_files_and_test_files_options/lib/ttnt/testtask.rb#L127-L131) call `TestToCodeMapping#select_code_files!` method ([implementation here](https://github.com/Genki-S/ttnt/blob/code_files_and_test_files_options/lib/ttnt/test_to_code_mapping.rb#L67-L76)).
I took this approach since `TestToCodeMapping` do not know the option `code_files` in anchoring task (since tests are running in separate processes). It might be possible to make `select_project_files` regard `code_files` option if we use DRb, but I wanted to make it simple.
But having 2 filtering processes (one for `select_project_files` which selects files under git repository and one for `select_code_files!` which selects files in `code_files` option) might be more complex... What do you think?

robin850 commented:
Hi there!

Sorry for the late answer! This is looking good ; as far as I can understand, there's no big deal with the `select_project_files!` thing, we can handle this later anyway!

Anyway, what do you think about [this diff](https://gist.github.com/robin850/11d3141a5dcb4537ef22) ? Basically we don't share methods with `Rake::TestTask` anymore, we just delegate missing methods to it through `method_missing`. This simplifies the `initialize` method and this allows these syntaxes (previously, the last one wouldn't work as expected and the second one was pretty limited):

~~~ruby
# Passing only a task instance
TTNT::TestTask.new(t)

# Passing only a block
TTNT::TestTask.new do |t|
  t.pattern = # ...
end

# Passing both
TTNT::TestTask.new(t) do |t|
  t.name = :overriden
end
~~~

Genki-S commented:
> Anyway, what do you think about this diff ?

It looks perfect! It makes the code more clean and versatile. Thank you very much for your suggestion!
I applied your diff (and I removed other `@rake_testtask`s from the code as well) in https://github.com/Genki-S/ttnt/commit/22262203a3cd3cab98d875cba8cff4c6a6dd5285 and wrote about it to the README in https://github.com/Genki-S/ttnt/commit/101f7589f43cbd9d9bfb1352cb385951899e5033.

> there's no big deal with the select_project_files! thing, we can handle this later anyway!

Thank you! I am relieved :smile:

I'm merging this, but please let me know if you notice something in the new 2 commits :pray:

robin850 commented:
Awesome! :heart:
